File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s)
    {
        int left = 0;
        int right = 0;
        int ans = 0;
        vector<int> hash(128, 0);

        while (right < s.size())
        {
            if (hash[s[right]] == 0)
            {
                hash[s[right]] += 1;
                right++;
            }
            else
            {
                ans = max(ans, right - left);
                hash[s[left]] -= 1;
                left++;
            }
        }
        if (left < right)
            ans = max(ans, right - left);
        return ans;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        m = defaultdict(int)
        l = 0
        r = 0

        res = 0
        while r < len(s):
            print(r)
            m[s[r]] += 1

            while m[s[r]] > 1:
                m[s[l]] -= 1
                l += 1
            
            res = max(res, r - l + 1)

            r += 1
        
        return res

